dask/dataframe/multi.py

```diff
@@ -308,20 +308,28 @@ def hash_join(lhs, left_on, rhs, right_on, how='inner',
 
 
 def single_partition_join(left, right, **kwargs):
+    # if the merge is perfomed on_index, divisions can be kept, otherwise the
+    # new index will not necessarily correspond the current divisions
+    get_divisions = lambda divisions, on_index: (
+        divisions if on_index else [None for _ in divisions]
+    )
+
     meta = pd.merge(left._meta_nonempty, right._meta_nonempty, **kwargs)
     name = 'merge-' + tokenize(left, right, **kwargs)
     if left.npartitions == 1:
         left_key = first(left._keys())
         dsk = dict(((name, i), (apply, pd.merge, [left_key, right_key],
                                 kwargs))
                    for i, right_key in enumerate(right._keys()))
-        divisions = right.divisions
+        divisions = get_divisions(right.divisions, kwargs.get('right_index'))
```

Comment:
I have no strong preference either. I used an if-statement to stay within the 80 characters.